File: 백준/Silver/2583. 영역 구하기/영역 구하기.py
import sys
sys.setrecursionlimit(10000)

M, N, K = map(int, input().split())
answer = []

position_list = []  # 사각형의 좌표들을 받기 위한 리스트
dfs_list = [[1 for i in range(N)] for k in range(M)]

for _ in range(K):
    position_list.append(list(map(int, input().split())))

for p in position_list:
    for x in range(p[0], p[2]):
        for y in range(p[1], p[3]):
            dfs_list[y][x] = 0 # 직사각형 내부의 좌표를 0으로 초기화

# 여기까지가 기본 초기화

def dfs(m, n):

    global count

    if m >= M or m < 0 or n >= N or n < 0:
        return False

    if dfs_list[m][n] == 0:
        return False

    dfs_list[m][n] = 0

    dfs(m+1, n)
    dfs(m-1, n)
    dfs(m, n+1)
    dfs(m, n-1)
    count += 1

    return count

for x in range(M):
    for y in range(N):
        count = 0
        response = dfs(x, y)
        if response != False:
            answer.append(response)

answer.sort()
print(len(answer))

for a in answer:
    print(a, end=" ")

File: 백준/Silver/2583. 영역 구하기/영역 구하기.js
const { dir } = require("console");
const fs = require("fs");
const stdin = (
  process.platform === "linux"
    ? fs.readFileSync("/dev/stdin").toString()
    : fs.readFileSync("../input.txt").toString()
).split("\n");

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

const [M, N, K] = input()
  .trim()
  .split(" ")
  .map((v) => Number(v));

const area = Array.from(Array(M), () => Array(N).fill(true));
const visited = Array.from(Array(M), () => Array(N).fill(false));

const direction = [
  [0, 1],
  [0, -1],
  [1, 0],
  [-1, 0],
];

const rangeList = [];

for (let i = 0; i < K; i++) {
  rangeList.push(
    input()
      .trim()
      .split(" ")
      .map((v) => parseInt(v))
  );
}

for (const r of rangeList) {
  for (let y = r[1]; y < r[3]; y++) {
    for (let x = r[0]; x < r[2]; x++) {
      area[y][x] = false;
    }
  }
}

// y + d[0] >= 0 && y + d[0] < N && x + d[1] >= 0 && x + d[1] < M;

function dfs(y, x, originY, originX) {
  for (const d of direction) {
    if (
      y + d[0] >= 0 &&
      y + d[0] < M &&
      x + d[1] >= 0 &&
      x + d[1] < N &&
      !visited[y + d[0]][x + d[1]] &&
      area[y + d[0]][x + d[1]]
    ) {
      visited[y + d[0]][x + d[1]] = true;
      area[originY][originX] += 1;
      dfs(y + d[0], x + d[1], originY, originX);
    }
  }
}

const answer = [];

for (let y = 0; y < M; y++) {
  for (let x = 0; x < N; x++) {
    if (area[y][x] && !visited[y][x]) {
      area[y][x] = 1;
      visited[y][x] = true;
      dfs(y, x, y, x);
      answer.push(area[y][x]);
    }
  }
}

console.log(answer.length);
answer.sort((x, y) => x - y);
console.log(answer.join(" "));